idanalysis/eqparams.py

```diff
@@ -0,0 +1,1294 @@
+"""IDs analysis."""
+
+import mathphys
+import matplotlib.pyplot as _plt
+import numpy as _np
+import pyaccel
+import pymodels
+from fieldmaptrack import Beam
+from scipy.integrate import cumtrapz
+from scipy.optimize import minimize
+
+ECHARGE = mathphys.constants.elementary_charge
+EMASS = mathphys.constants.electron_mass
+LSPEED = mathphys.constants.light_speed
+VCPERM = mathphys.constants.vacuum_permitticity
+CQ = mathphys.constants.Cq
+ECHARGE_MC = ECHARGE / (EMASS * LSPEED)
+
+_model = pymodels.si.create_accelerator()
+_mid_subsections = None
+
+
+class InsertionParams:
+    """Class to specify insertion parameters.
+
+     All equations used here are from the book: Clarke, James A.,
+    'The Science and Technology of Undulators and Wigglers'.
+    Chapter 10: 'The Effect of Insertion Devices on the Electron Beam'.
+    """
+
+    def __init__(self, beam_energy=3.0):
+        """Class constructor.
+
+        Args:
+            beam_energy (Float, optional): Beam energy [GeV]. Defaults to 3.0.
+        """
+        self._field_profile = None
+        self._fam_name = None
+        self._bx_peak = None
+        self._by_peak = None
+        self._kx = None
+        self._ky = None
+        self._period = None
+        self._length = None
+        self._nr_periods = None
+        self._straight_section = None
+        self._etax = None
+        self._etapx = None
+        self._etay = None
+        self._etapy = None
+        self._i1x = None
+        self._i2x = None
+        self._i3x = None
+        self._i4x = None
+        self._i5x = None
+        self._i1y = None
+        self._i2y = None
+        self._i3y = None
+        self._i4y = None
+        self._i5y = None
+        self._u0 = None
+        self._model = None
+        self._mid_subsections = None
+        self.beam = Beam(beam_energy)
+        self.brho = self.beam.brho
+        self.gamma = self.beam.gamma
+
+    @property
+    def model(self):
+        """Sirius model.
+
+        Returns:
+            Pymodels object: Si model
+        """
+        return self._model
+
+    @model.setter
+    def model(self, value):
+        global _model
+        global _mid_subsections
+        global _twiss
+        _model = value
+        mia = pyaccel.lattice.find_indices(_model, "fam_name", "mia")
+        mib = pyaccel.lattice.find_indices(_model, "fam_name", "mib")
+        mip = pyaccel.lattice.find_indices(_model, "fam_name", "mip")
+        _mid_subsections = _np.sort(_np.array(mia + mib + mip))
+        _twiss, *_ = pyaccel.optics.calc_twiss(_model, indices="open")
+        self._model = _model
```

Comment:
why do you need to keep a reference of the model inside the object? This could be misleading, couldn't?

Reply:
we talked offline about this issue.